### PR TITLE
adding support for XML parsing and shale type mapping

### LINKED ISSUE #47: Plan adding  support for XML parsing and Shale-type mapping.
Agree with the enhancement. However, I think we should probably collect these enhancements to Shale in another location, and instead of a mix-and-match of using Shale class vs this overridden class, we should only use the overridden class.

In fact, I want to enhance our LutaML gem to support Ruby object access such as Shale does. LutaML already provides a way to define a data model, we just need to extend it to support XML parsing and Shale-type mapping.

The conclusion is I'd like to update the LutaML gem to contain this Shale hack... @HassanAkbar can you help plan this feature at LutaML?

_Originally posted by @ronaldtse in https://github.com/metanorma/sts-ruby/pull/7#discussion_r1460256133_

### PR DESCRIPTION
**Pull request by HassanAkbar**
Adding support for XML parsing and Shale-type mapping.

Fixes #47

### CODE REVIEW

**Comment by HassanAkbar** [comment]
@ronaldtse I've tried to make the structure for parsing XML as close to other structures as possible. is this correct or do we need to change something here?

**Comment by ronaldtse** [comment]
@HassanAkbar actually this is not quite what I wanted. We need a way to allow users to "access XML element and attributes".

So the user should provide an XML schema, and also an XML document, like this, and allow the user to access elements:
* https://www.shalerb.org/#compiling-json-and-xml-schema

**Comment by HassanAkbar** [comment]
@ronaldtse I don't fully understand what you mean, But I will look into it and get back to you with questions related to this after some research.

**Comment by HassanAkbar** [comment]
@ronaldtse I've looked at https://www.shalerb.org/#compiling-json-and-xml-schema, and figured how `shale` gem can generate ruby code from schemas provided in `xml` or `json` format. 

From what I understand the flow is something like, 
- User provides an `xml document`. (this point is clear)
- User provides the `xml schema`. (this is unclear, I have few questions related to this)
  - How will the user provide the schema? will he pass it as a string parameter or as a file or folder name?
  - Are we going to support both `xml` and `json` schemas or only `xml` schemas?
- We will return the shale object for the given `xml_document`.

**Comment by HassanAkbar** [comment]
@ronaldtse I have updated this PR. Is this what you were looking for ?

File: lib/lutaml/xml/parsers/xml.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    module Parsers
+      # Class for parsing .xml schema files into ::Lutaml::Uml::Document
+      class Xml
+        def self.load_schema(schema, root_schema)
+          result = Shale::Schema.from_xml([schema])
+
+          result.values.each do |klass|
+            # Temporary solution will update in the parser
+            klass = klass.gsub(/^require.*?\n/, "")
+            klass = klass.gsub(/< Shale::Mapper/, "< Lutaml::Xml::Mapper")
```
Comment (ronaldtse) [inline lib/lutaml/xml/parsers/xml.rb:19]: We need to fork the gem

Comment (HassanAkbar) [reply lib/lutaml/xml/parsers/xml.rb:19]: @ronaldtse Sure, I'll fork this and open a PR in that repo.

Comment (HassanAkbar) [reply lib/lutaml/xml/parsers/xml.rb:19]: @ronaldtse Forked the gem at lutaml/shale -> https://github.com/lutaml/shale

**Review by ronaldtse — APPROVED** [review]
Merging first.

**Pull request merged**